packages/graph/hash_graph/lib/graph/src/datastore/mod.rs

```diff
@@ -0,0 +1,31 @@
+use std::fmt;
+use error_stack::{Result, Context};
+
+mod postgres;
+
+pub use postgres::PostgresDatabase;
+use crate::types::{GraphElementIdentifier, GraphElementType};
+
+#[derive(Debug)]
+pub struct DatastoreError;
+
+impl Context for DatastoreError {}
+
+impl fmt::Display for DatastoreError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.write_str("The Datastore encountered an error")
+    }
+}
+
+/// Describes the API of a Datastore implementation
+trait Datastore {
+    // TODO - Perhaps we want explicit `create_entity_type` methods and such rather than trying to
+    //  make this general
```

Comment:
This is one of the things I'm most unsure about.

Perhaps we do just want all of the methods like `create_entity_type`, `create_entity`, etc. Generalising it might not have much worth.

I was also thinking we could explore generics, associated types, or possibly some other grouping where we have something like `EntityType::update(dbPool: PostgresPool)`, but right now I'm leaning to that last suggestion being pretty gross. 

@TimDiekmann @thehabbos007 not sure if you have thoughts on how we want to structure it at this point, or if we just want to go with it as is and see how it evolves

Reply:
I have some ideas on solving this by using generics. Will take a proper look tomorrow.